src/exp2.jl

```diff
@@ -0,0 +1,90 @@
+# Import the generated code
+for i=1:15
+    include("exp_generated/exp_$i.jl")
+end
+
+function getmem(cache,k)
+    return cache[k-1];
+end
+
+function _exp2!(A; caches=nothing, do_balancing = A isa StridedMatrix)
+    n = LinearAlgebra.checksquare(A)
+    nA = opnorm(A,1);
+
+    # Maybe to balancing
+    if do_balancing
+        ilo, ihi, scale = LAPACK.gebal!('B', A)    # modifies A
+    end
+
+    # Select how many multiplications to use
+    rhov=    [0.015; 0.25; 0.95; 2.1; 5.4];
+    # Number of memslots needed (beside A)
+    memslots=[3   ;    4;    5;   6;   5];
+    for s=1:7 # Only 8 since exp(5.4*2^8)=Inf
+        push!(rhov, rhov[end]*2);
+        push!(memslots,memslots[end]);
+    end
+    i = findfirst(nA < rho for rho in rhov)
+
+    # Take care of caching
+    if caches == nothing
+        # No cache. Allocate Matrix objects
+        TT=eltype(A);
+        thiscache=[Matrix{TT}(undef,n,n) for i=1:memslots[i]];
+    else
+        thiscache = caches;
+    end
+
+
+    # Evaluate
+    if (nA < rhov[1])
```

Comment:
Could make it a `Val` arg, and then something like
```julia
Base.Cartesian.@nif 13 d -> nA < rhov[d] d -> X = __exp!(thiscache, A, Val(d)) d -> X = __exp!(thiscache, A, Val(d))
```
This produces
```julia
:(if nA < rhov[1]
      X = __exp!(thiscache, A, Val(1))
  else
      if nA < rhov[2]
          X = __exp!(thiscache, A, Val(2))
      else
          if nA < rhov[3]
              X = __exp!(thiscache, A, Val(3))
          else
              if nA < rhov[4]
                  X = __exp!(thiscache, A, Val(4))
              else
                  if nA < rhov[5]
                      X = __exp!(thiscache, A, Val(5))
                  else
                      if nA < rhov[6]
                          X = __exp!(thiscache, A, Val(6))
                      else
                          if nA < rhov[7]
                              X = __exp!(thiscache, A, Val(7))
                          else
                              if nA < rhov[8]
                                  X = __exp!(thiscache, A, Val(8))
                              else
                                  if nA < rhov[9]
                                      X = __exp!(thiscache, A, Val(9))
                                  else
                                      if nA < rhov[10]
                                          X = __exp!(thiscache, A, Val(10))
                                      else
                                          if nA < rhov[11]
                                              X = __exp!(thiscache, A, Val(11))
                                          else
                                              if nA < rhov[12]
                                                  X = __exp!(thiscache, A, Val(12))
                                              else
                                                  X = __exp!(thiscache, A, Val(13))
                                              end
                                          end
                                      end
                                  end
                              end
                          end
                      end
                  end
              end
          end
      end
  end)
```

Reply:
And then add a function `__exp!`? Looks quite complicated. I'm slightly in favor of the current hard-coded if-statements.

Reply:
It'd be replacing `exp12(x,y) = ...` with `__exp(x, y, ::Val{12}) = ...`.

Reply:
Okay. I understand. Rather than many different functions with different names, one could selected it with dispatch. It's doable with the code generation. Advantages?

Reply:
Mostly that it's fewer lines of code, and easier to cover with tests.
I agree that it is questionable whether it's actually easier to read.
It can be broken up to make it a little more clear:
```julia
X = Base.Cartesian.@nif 13 d -> begin
    nA < rhov[d]
end d -> begin # if condition
    __exp!(thiscache, A, Val(d))
end d -> begin # fallback (d == 13)
    __exp!(thiscache, A, Val(d))
end
```

Reply:
That looks okay to read. Done.